### PR TITLE
Add initial implementation

### PR DESCRIPTION
**Pull request by luislee818**
This is a port of existing `apiSmith` from HTML-Client.

### CODE REVIEW

**Comment by luislee818** [comment]
@FMRb @kasperlewau This is the first time I create an NPM module, any suggestion is welcome. Thanks in advance!

File: .babelrc [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+  "presets": ["es2015"]
```
Comment (kasperlewau) [inline .babelrc]: I would use `babel-preset-env` instead so as to only transpile what needs to be transpiled for the desired browser matrix 😄

File: .gitignore [hunk 2]
```diff
@@ -0,0 +1,2 @@
+node_modules
+dist
```
Comment (kasperlewau) [inline .gitignore]: add npm-debug.log

File: package.json [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  "description": "A lightweight library for declarative object transformations in JavaScript.",
+  "homepage": "https://github.com/FalconSocial/js-api-smith",
+  "repository": {
+    "type": "git",
+    "url": "git://github.com/FalconSocial/js-api-smith.git"
+  },
+  "main": "dist/index.js",
+  "scripts": {
+    "test": "mocha --require babel-core/register --reporter spec",
+    "test:w": "mocha --require babel-core/register --watch --reporter spec",
```
Comment (kasperlewau) [inline package.json]: For a JS library that doesn't depend on the browser to run, I would run tests inside of `tape` or `ava`. Or any other minimalist test runner - Mocha is quite the monolith by comparison. 

`tape` being the preferred option as it has more widespread use, and it's a hell of a lot more stable than `ava` :)

**Comment by luislee818** [comment]
@kasperlewau thanks for the suggestions, I've added `babel-preset-env` and started using `tape` (it's pretty different 😄 ). I will take look at the _drone_ integration later.

I hope I've done them correctly, if you see anything else can be improved, please let me know.

**Comment by luislee818** [comment]
Mr. Smith would be happy to be free. 🏡

File: package.json [hunk 4]
```diff
@@ -24,18 +24,34 @@
...
+  },
+  "babel": {
+    "presets": [
+      [
+        "env",
+        {
+          "targets": {
+            "browsers": [
+              "last 2 versions",
+              "ie >= 7"
```
Comment (kasperlewau) [inline package.json]: 7!? **Dayumn son**

**Comment by kasperlewau** [comment]
> I hope I've done them correctly, if you see anything else can be improved, please let me know.

I'll take a more in-depth look at the tests on my way home / sometime tonight I reckon. Nicely done! 👍

**Comment by kasperlewau** [comment]
Here's a thought: 

Let's add this to `package.json`: 

```json
"browserify": {
  "transform": [
    "babelify"
  ]
}
```

And then we'll set `"main"` to be `src/index.js` and drop the whole `build` process in this repo. 

That will make for an easier life requiring/importing the file from HTML-Client (https://github.com/FalconSocial/HTML-Client/pull/4147)

**Comment by luislee818** [comment]
After adding the _babelify_ transformation in _browserify_ inside `package.json`, and removing the `build` task, I can publish the node module without problem.

However, when I try to do `gulp dev` in HTML-Client (after `npm install`), I got the following error:
```
✖ Error: Couldn't find preset "env" relative to directory "/Users/dapeng/Workspace/HTML-Client/node_modules/@falconio/js-api-smith" while parsing file: /Users/dapeng/Workspace/HTML-Client/node_modules/@falconio/js-api-smith/src/index.js
```
I guess it's because we don't use `babel-preset-env` in HTML-Client? For minimal change I suppose I could use `es2015` babel preset in my module repo?

If I understand correctly, the build/transpilation process couples the consumer (HTML-Client) and the dependency (this node module), in the module we specify the _transformation_ and _transformation options_, but that relies on the consumer knows and supports them, right? 😄

**Comment by kasperlewau** [comment]
Hmm. 

I think there's two ways to correct this: 

1. Make the `babel-preset-env` a regular `dependency` so as to install it together with `js-api-smith`. 
  * Not particularly fond of this one. 
2. Switch to `babel-preset-env` in the HTML-Client and see if that does it.
  * If this works by itself, we're golden. 

> If I understand correctly, the build/transpilation process couples the consumer (HTML-Client) and the dependency (this node module), in the module we specify the transformation and transformation options, but that relies on the consumer knows and supports them, right?

When we run `browserify`, it has an idea of our own code and third party code. Third party code is anything that is imported _from_ `node_modules`. 

When we want to apply a _transform_ to our browserify process, we apply it to our own code - _not_ our third party code. 

However, some third party packages might need transformations applied to them (`js-api-smith`, for example) and should specify that in their `package.json` (just like you've done now). 

In doing so, `browserify` wont ignore the package even though it is regarded a third party module. 

Another (**naive**) approach would be to run our `babelify` transformation with the `-g/--global` flag, but that would run _every third party module through `babel`._ Definitely not what we want. 

--- 

As far as module/package resolution goes, I _think_ `browserify` will start at the location of the dependency and traverse upwards through the tree. 

```
# with babel-preset-env as a _dependency_
# npm install babel-preset-env --save

└── node_modules
    ├── babel-preset-es2015
    ├── babelify
    ├── browserify
    └── js-api-smith
        └── node_modules
            └── babel-preset-env # <- found babel-preset-env relative to `/path/to/js-api-smith`
```

```
# _without_ babel-preset-env as a dependency

└── node_modules
    ├── babel-preset-es2015
    ├── babelify
    ├── browserify
    └── js-api-smith
        └── node_modules # <- babel-preset-env not found relative to `/path/to/js-api-smith`
```

```
# with babel-preset-env as a dependency in the consumer 

└── node_modules
    ├── babel-preset-env # <- babel-preset-env found in parent
    ├── babelify
    ├── browserify
    └── js-api-smith
        └── node_modules 
```

This is all very much speculation and trying to remember the things I've read over the years, but I think the most straight-forward solution is to tie-in and switch to `babel-preset-env` all across our repositories. 

It is after all, the best preset we can use from a browser matrix/support perspective

**Comment by luislee818** [comment]
@kasperlewau Thanks so much for the detailed explanation, after spending a couple of days looking into the build process I found more intricacies and your answer definitely helped me making sense of some of them. 🙇 

BTW, your second suggestion of adding `babel-preset-env` to HTML-Client seems to work, I will create a sandbox and try it out. 😄

**Pull request merged**